### PR TITLE
Fix InvalidIpException due to stripping (issue #16992)

### PR DESCRIPTION
**Pull request by wdoekes**
Provides a possible fix to #16992. By wrapping SnmpQuery, we get proper values for all IP addresses and won't need to throw InvalidIpExceptions.

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.
- [ ] If my Pull Request makes discovery/polling/yaml changes, I have added/updated [test data](https://docs.librenms.org/Developing/os/Test-Units/).

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

File: includes/snmp.inc.php [hunk 1]
```diff
@@ -399,7 +400,59 @@ function snmp_walk($device, $oid, $options = null, $mib = null, $mibdir = null)
...
+            $new_result[$key_num] = [];
+        }
+
+        $new_result[$key_num][$key] = $value;
+    }
+
+    return $new_result;
+}//end snmpquery_result_to_snmpwalk_result()
+
 function snmpwalk_cache_oid($device, $oid, $array = [], $mib = null, $mibdir = null, $snmpflags = '-OQUs')
```
Comment (murrant) [inline includes/snmp.inc.php:436]: There are a lot of corner cases.  Something like this should probably be a separate PR as it is a big change.

Is the IP::fromHexString() change enough to fix your issue?  If so, I'd suggest dropping the snmpwalk changes from this PR as it will be easier to get merged.

Comment (wdoekes) [reply includes/snmp.inc.php:436]: > Is the IP::fromHexString() change enough to fix your issue?

It is not. I don't have spaces or double-quotes in the IPs. I simply noticed that that code was also wrong.

I do have backslashes in the IPs and they need the bigger fix.

> There are a lot of corner cases.

I did spend the day checking for corner cases. By comparing the counts of results between this version and the original one.

And the code change itself is trivial to read. If you trust SnmpQuery then this new code should be better.

If it always returns an array of `"$MIB::$OID[...][...]" => $value` then it should be good to go.

Comment (murrant) [reply includes/snmp.inc.php:436]: I think it would be more effective to convert this one use to SnmpQuery.

Comment (wdoekes) [reply includes/snmp.inc.php:436]: So, the CI reports this:

```
--- Expected
+++ Actual
@@ @@
     'ports.1.ifName' => 'eth0'
     'ports.1.portName' => null
     'ports.1.ifIndex' => 2
-    'ports.1.ifSpeed' => 1000000000
+    'ports.1.ifSpeed' => null
...
     'ports.1.ifType' => 'ethernetCsmacd'
     'ports.1.ifAlias' => 'eth0'
-    'ports.1.ifPhysAddress' => '1ae8291e393b'
+    'ports.1.ifPhysAddress' => null
```
What gives? Does NetSnmpQuery require more flags?

Comment (wdoekes) [reply includes/snmp.inc.php:436]: > I think it would be more effective to convert this one use to SnmpQuery.

Effective, as in to move this forward. Yes. But then the next guy with an IP address in a different MIB runs into that problem as well. And then _they_ wonder why there are two competing snmp_walks where one is clearly broken.

Comment (murrant) [reply includes/snmp.inc.php:436]: I'm the next guy ;)  Been working to convert snmp queries.
One less query to convert will save me time.

Most MIBs return properly formatted IPs instead of hex nonsense ;)  So this isn't a problem for everything.

Comment (wdoekes) [reply includes/snmp.inc.php:436]: Hm. As you can see [here](https://github.com/librenms/librenms/issues/16992#issuecomment-2598668628), this patch already has positive effects elsewhere. I'm definitely open to correcting the codebase with regards to replacing old library calls. But not getting automatic improvements by fixing internals feels eternally wrong.
```
# find includes/ LibreNMS/ -name '*.php' | xargs grep -F 'snmpwalk_cache_oid(' | wc -l
420
```

Comment (laf) [reply includes/snmp.inc.php:436]: It's really appreciated that you want to help update the codebase but you probably need to cut us some slack here, no one is trying to stop things being improved but we see the impact small changes have sometimes and when something breaks a corner case not covered by tests then it's (mostly murrant) who ends up hot fixing stuff.

As you can see already, tests are now broken and that's the parts we know about.

Comment (murrant) [reply includes/snmp.inc.php:436]:
```
 find includes/ LibreNMS/ -name '*.php' | xargs grep -F 'snmpwalk_cache_oid(' | wc -l
411
```

9 down, 411 to go :)

Comment (murrant) [reply includes/snmp.inc.php:436]: It does look like your change broke at least ifSpeed for port polling in at least some cases.

Comment (wdoekes) [reply includes/snmp.inc.php:436]: That's quite interesting actually. Now I'm intrigued. Can't promise when I have time to work on this though.

Good to know that there are some conversion examples (like https://github.com/librenms/librenms/commit/3b8ac6f9e18fc6d21ed0cd6c8d01919da3f6d7b5) to steal from.

Comment (wdoekes) [reply includes/snmp.inc.php:436]: In the mean time, here's an alternative quick fix:
```diff
diff --git a/includes/polling/cipsec-tunnels.inc.php b/includes/polling/cipsec-tunnels.inc.php
index 85ac302c3..b492aa28e 100644
--- a/includes/polling/cipsec-tunnels.inc.php
+++ b/includes/polling/cipsec-tunnels.inc.php
@@ -5,9 +5,9 @@ use LibreNMS\Util\IP;
 
 if ($device['os_group'] == 'cisco') {
     // FIXME - seems to be broken. IPs appear with leading zeroes.
-    $ipsec_array = snmpwalk_cache_oid($device, 'cipSecTunnelEntry', [], 'CISCO-IPSEC-FLOW-MONITOR-MIB');
+    $ipsec_array = snmpwalk_cache_oid($device, 'cipSecTunnelEntry', [], 'CISCO-IPSEC-FLOW-MONITOR-MIB', null, $snmpflags = '-OQUsx');
     if (! empty($ipsec_array)) {
-        $ike_array = snmpwalk_cache_oid($device, 'cikeTunnelEntry', [], 'CISCO-IPSEC-FLOW-MONITOR-MIB');
+        $ike_array = snmpwalk_cache_oid($device, 'cikeTunnelEntry', [], 'CISCO-IPSEC-FLOW-MONITOR-MIB', null, $snmpflags = '-OQUsx');
     }
 
     $tunnels_db = dbFetchRows('SELECT * FROM `ipsec_tunnels` WHERE `device_id` = ?', [$device['device_id']]);
```

**Comment by murrant** [comment]
Closing to do lack of updates additionally, I think this is the wrong way to fix this.
